### PR TITLE
provide access to stats for system health monitoring via xymon

### PR DESCRIPTION
**Pull request by caustin-usgs**

### CODE REVIEW

**Comment by jakebrinkmann** [comment]
Thinking the following should become accessible: 
```
/production-api/v1/statistics/stat_open_orders
/production-api/v1/statistics/stat_waiting_users
/production-api/v1/statistics/stat_backlog_depth
/production-api/v1/statistics/stat_products_complete_24_hrs
/production-api/v1/statistics/stat_onorder_depth
```

**Comment by caustin-usgs** [comment]
yes, they would. Bills aim though was to access stat_products_complete_24_hrs and stat_onorder_depth

**Comment by jakebrinkmann** [comment]
Ok, I had to add admin_statlist=xxx.xxx.xxx.xxx to my config file (maybe there should be a check if `white_ls` is not None 😄 ), seems to be working when pointed at the devDB

**Comment by jakebrinkmann** [comment]
Seems a lot of the code is very similar to the `admin_whitelist`
[production-api/v0/statistics](https://github.com/USGS-EROS/espa-api/blob/59e42e5e34dad03dd8f5887ea1a4ec8fd6be606a/api/transports/http.py#L99) seems same as [api/v0/statistics](https://github.com/USGS-EROS/espa-api/blob/master/api/transports/http.py#L74)
[ProductionStats](https://github.com/USGS-EROS/espa-api/blob/59e42e5e34dad03dd8f5887ea1a4ec8fd6be606a/api/transports/http_admin.py#L135) seems same as [Reports](https://github.com/USGS-EROS/espa-api/blob/master/api/transports/http_admin.py#L96)
[stats_whitelist](https://github.com/USGS-EROS/espa-api/blob/59e42e5e34dad03dd8f5887ea1a4ec8fd6be606a/api/transports/http_admin.py#L47) seems same as [whitelist](https://github.com/USGS-EROS/espa-api/blob/master/api/transports/http_admin.py#L25)

Would those similar functions be removed from /api, since they have been moved to /production-api ?
I would imagine adding the same [espa.available_stats()](https://github.com/USGS-EROS/espa-api/blob/master/api/transports/http_admin.py#L113) would be a desired addition, too?

---
I am running `nose2` now, though I don't see any specific tests for this functionality

**Comment by caustin-usgs** [comment]
it is, b/c they're decorators for different functions, serving different ends.  the new 'stats_whitelist' decorator protects a new production-api endpoint, meant solely for xymon to be able to retrieve those returned values without us creating a faux user.  we can't use the admin_whitelist decorator since that is used to verify that only our workers can access those endpoints

**Comment by jakebrinkmann** [comment]
Hey Clay, I was checking out how to add a test for this, and had a question:

The API has a variety of endpoints, take these three for example:

1. `/production-api/v1/set_product_unavailable`
1. `/api/v1/statistics/`
1. `/production-api/v1/statistics`

Seems 1 is for the hadoop workers only, 2 is for ESPA-WEB (authenticated users, or an admin_whitelist), and 3 is for the new xymon interface. 

Now, when these three are "whitelisted" it seems 2&3 are configured from the config file (`~/.usgs/.cfgnfo`), while the whitelist for 1 is taken from the database. 

If I am understanding this correctly, is there a reason for this difference? Could they all theoretically be taken from the database?

I ask, because when spoofing an IP in the tests, I could quite easily emulate [the same tests as for the hadoop workers](https://github.com/USGS-EROS/espa-api/blob/master/test/test_flask_production_transport.py#L61), where with the setup right now, I would have to think a bit more about what I am spoofing.... As it seems there are many new possibilities on *what* to spoof/mock

**Comment by jakebrinkmann** [comment]
Hmmm, I am afraid I am inclined to suggest we change back to [how the authentication was before](https://github.com/USGS-EROS/espa-api/pull/46/commits/b3b3ef60bb232801904e0aabca6e87278368e689#diff-8c3fadb895bfaf2f3fdb38ef13638e62L167)

This way, patch can override that class method (get_admin_whitelist) with a spoofed version. I am having issues patching over the [decorator here](https://github.com/USGS-EROS/espa-api/pull/46/commits/b3b3ef60bb232801904e0aabca6e87278368e689#diff-cbcd1ca0c6c98a05c651e072b5f671a6R33)

Otherwise it seems the functionality **requires** the `admin_statlist=xxx.xxx.xxx.xxx` to already be inside the config file, but that would require localhost (127.0.0.1) to exist in every config file... unless I can figure out how to override that decorator 

I've tried defining a mock get_stats_whitelist inside [api.interfaces.ordering.mocks.version1](https://github.com/USGS-EROS/espa-api/blob/master/api/interfaces/ordering/mocks/version1.py#L18) and using this patch (or variations) to no avail: 

```python
    @patch('api.transports.http_admin.ProductionStats.decorators', [api.get_stats_whitelist])
```

I just can't seem to override anything in `api.transports.http_admin`

**Comment by jakebrinkmann** [comment]
Hey Clay, with these changes, I think (if I didn't break anything) it should still behave the way it did with your changes, but allow easier addition of more "whitelisted" functionality unittests. 
Note: I changed `admin_statlist` in the config file to read `stats_whitelist`, seemed to fit better with other naming conventions
Cheers 🍻 
Jake

File: api/interfaces/admin/version1.py [hunk 1]
```diff
@@ -164,6 +164,30 @@ def get_stat(self, name):
...
+        try:
+            response = self.admin.admin_whitelist()
+        except:
+            logger.debug("ERR failure to generate production whitelist\ntrace:{}".format(traceback.format_exc()))
+            response = default_error_message
+        return response
+
+    def get_stat_whitelist(self):
+        """
+        Returns list of ip addresses in xymon cluster
```
Comment (caustin-usgs) [inline api/interfaces/admin/version1.py]: xymon is not a cluster. its just a monitoring application

File: api/providers/administration/administration_provider.py [hunk 2]
```diff
@@ -126,4 +127,8 @@ def get_system_config():
 
     @staticmethod
     def admin_whitelist():
-        return ['127.0.0.1']
\ No newline at end of file
+        return api_cfg()['admin_whitelist']
+
```
Comment (caustin-usgs) [inline api/providers/administration/administration_provider.py]: make sure both admin_whitelist and stat_whitelist are defined in all 3 environments config files before deploying.  no harm in having them there early, but dont remove one that may still be in use

**Comment by caustin-usgs** [comment]
good points @jakebrinkmann feel free to merge when you're ready

**Pull request merged**